src/libraries/Microsoft.PowerFx.Core/Texl/Builtins/Remove.cs

```diff
@@ -0,0 +1,579 @@
+
+            var args = Contracts.VerifyValue(callNode.Args.Children);
+
+            DType dsType = binding.GetType(args[0]);
+            if (dsType.AssociatedDataSources == null
+                || dsType == DType.EmptyTable)
+            {
+                return false;
+            }
+
+            var sourceRecordType = binding.GetType(args[1]);
+
+            // This might be the case where Remove(CDS, Gallery.Selected)
+            if (sourceRecordType == DType.EmptyRecord)
+            {
+                return false;
+            }
+
+            var firstTypeName = sourceRecordType.GetNames(DPath.Root).FirstOrDefault();
+
+            if (!firstTypeName.IsValid)
+            {
+                return false;
+            }
+
+            DType type = firstTypeName.Type;
+            DName columnName = firstTypeName.Name;
+
+            // This might be the case where Remove(CDS, Gallery.Selected)
+            if (!dsType.Contains(columnName))
+            {
+                return false;
+            }
+
+            dsType.AssociateDataSourcesToSelect(
+                dataSourceToQueryOptionsMap,
+                columnName,
+                type,
+                false /*skipIfNotInSchema*/,
+                true); /*skipExpands*/
+
+            return true;
+        }
+
+        // This method filters for a table as the first parameter, records as intermediate parameters
+        // and a string (First/All) as the final parameter.
+        public override bool IsSuggestionTypeValid(int paramIndex, DType type)
+        {
+            Contracts.Assert(paramIndex >= 0);
+            Contracts.AssertValid(type);
+
+            if (paramIndex >= MaxArity)
+            {
+                return false;
+            }
+
+            if (paramIndex == 0)
+            {
+                return type.IsTable;
+            }
+
+            // String suggestions for column names may occur within the context of a record
+            return type.IsRecord || type.Kind == DKind.String;
+        }
+
+        public override bool IsAsyncInvocation(CallNode callNode, TexlBinding binding)
+        {
+            Contracts.AssertValue(callNode);
+            Contracts.AssertValue(binding);
+
+            return Arg0RequiresAsync(callNode, binding);
+        }
+
+        protected override bool RequiresPagedDataForParamCore(TexlNode[] args, int paramIndex, TexlBinding binding)
+        {
+            Contracts.AssertValue(args);
+            Contracts.AssertAllValues(args);
+            Contracts.Assert(paramIndex >= 0 && paramIndex < args.Length);
+            Contracts.AssertValue(binding);
+            Contracts.Assert(binding.IsPageable(Contracts.VerifyValue(args[paramIndex])));
+
+            // For the first argument, we need only metadata. No actual data from datasource is required.
+            return paramIndex > 0;
+        }
+    }
+
+    // Remove(collection:*[], source:*[], ["All"])
+    internal class RemoveAllFunction : BuiltinFunction
```

Comment:
>RemoveAllFunction [](http://example.com/codeflow?start=19&length=17)

This function looks very similar to the previous one; can we have a single one?

Reply:
Ah, I noticed the difference (record arguments vs. table arguments). It can probably be refactored, though, to a common base type, so that most common code can be written once.